src/python/pants/bsp/goal.py

```diff
@@ -0,0 +1,47 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import logging
+import os
+import sys
+
+from pants.base.exiter import ExitCode
+from pants.base.specs import Specs
+from pants.bsp.protocol import BSPConnection
+from pants.build_graph.build_configuration import BuildConfiguration
+from pants.engine.unions import UnionMembership
+from pants.goal.builtin_goal import BuiltinGoal
+from pants.init.engine_initializer import GraphSession
+from pants.option.options import Options
+
+_logger = logging.getLogger(__name__)
+
+
+class BSPGoal(BuiltinGoal):
+    name = "bsp"
+    help = 'Run server for Build Server Protocol ("BSP").'
+
+    def run(
+        self,
+        *,
+        build_config: BuildConfiguration,
+        graph_session: GraphSession,
```

Comment:
Before using this `Session`, you'll likely want to "clone" it via `new_session(.., dynamic_ui=False)`, to ensure that requests to `Scheduler.execute` don't also steal your file handles (when in a TTY).

Reply:
The BSP goal replaces `sys.stdin` and `sys.stdout` with bytes-mode, unbuffered file objects. Can the `Session` or `Scheduler` interfere with those?

Reply:
fixed to use `new_session`

Reply:
> The BSP goal replaces sys.stdin and sys.stdout with bytes-mode, unbuffered file objects. Can the Session or Scheduler interfere with those?

Yes, for two reasons:
1. `sys.stdin` and `sys.stdout` have already been replaced with filehandles proxied to a socket
2. the UI will read/write to those same filehandles if it runs

Reply:
Is there a way for the BSP goal to take those over completely?

Reply:
No, but it doesn't need to. Nothing else should consume them aside from the UI... and actually, I think that the UI always runs on stderr anyway? So maybe this is moot.

Reply:
I will leave as-is then, and re-visit if and when there is a crossing of the streams (which hopefully does not cause total protonic reversal).